nrf_modem: v2.3.0

lemrey opened the pull request:
See CHANGELOG file for details.

eivindj-nordic commented on nrf_modem/include/nrf_modem_bootloader.h:
```diff
@@ -0,0 +1,154 @@
+ * @retval -NRF_ETIMEDOUT When modem did not respond.
+ * @retval -NRF_EIO When incorrect response received from modem.
+ */
+int nrf_modem_bootloader_update(void);
```
This signature is changed from `_apply` to `_update`. It should be mentioned in the changelog :)

eivindj-nordic commented on nrf_modem/include/nrf_socket.h:
```diff
@@ -805,6 +862,29 @@ int nrf_getaddrinfo(const char *restrict nodename,
+int nrf_getifaddrs(struct nrf_ifaddrs **ifa);
+
+/**
+ * @brief Free address information returned by @ref nrf_getaddrinfo().
```
nit:
```suggestion
 * @brief Free address information returned by @ref nrf_getifaddrs().
```

eivindj-nordic commented on nrf_modem/doc/CHANGELOG.rst:
```diff
@@ -9,6 +9,41 @@ Changelog
+
+Bootloader
+==========
+* The Full DFU API (:file:`nrf_modem_full_dfu.h`) has been moved to (:file:`nrf_modem_bootloader.h`) and renamed accordingly.
```
Maybe add a line about prefixing the DFU results with `NRF_`? E.g. `NRF_MODEM_DFU_RESULT_OK`

eivindj-nordic replied:
Also, some of the function signatures has changed the parameter order: (len, buf) -> (buf, len).

eivindj-nordic commented on nrf_modem/include/nrf_modem_os.h:
```diff
@@ -107,9 +107,20 @@ int32_t nrf_modem_os_timedwait(uint32_t context, int32_t *timeout);
  * @brief Notify the application that an event has occurred.
  *
+ * @param[in] context A identifier assigned by the library to identify the context of the
+ *                    event. Only wake threads with the same context.
```
We should add a note that if `context` is `0` all threads should wake up.

umapraseeda commented on nrf_modem/doc/architecture.rst:
```diff
@@ -19,22 +19,22 @@ The following figure shows a simplified Modem library architecture:
+The shared memory between the application and modem cores must reside within the lowest 128 kilobytes of RAM.
+The application is responsible for reserving the memory area before passing it onto the library.
+
+In bootloader mode, the shared memory area must be as large as ``NRF_MODEM_SHMEM_BOOTLOADER_SIZE`` and its base address must be 4-bytes aligned.
```
```suggestion
In bootloader mode, the shared memory area must be as large as the value of ``NRF_MODEM_SHMEM_BOOTLOADER_SIZE`` and its base address must be 4-bytes aligned.
```

umapraseeda commented on nrf_modem/doc/architecture.rst:
```diff
@@ -19,22 +19,22 @@ The following figure shows a simplified Modem library architecture:
+In bootloader mode, the shared memory area must be as large as ``NRF_MODEM_SHMEM_BOOTLOADER_SIZE`` and its base address must be 4-bytes aligned.
+The library accepts the shared memory base address and size as parameters to the :c:func:`nrf_modem_bootloader_init` function through the :c:struct:`nrf_modem_bootloader_init_params`.
+
+In normal operating mode the shared memory is divided into the following four regions:
```
```suggestion
In normal operating mode, the shared memory is divided into the following four regions:
```

umapraseeda commented on nrf_modem/doc/architecture.rst:
```diff
@@ -19,22 +19,22 @@ The following figure shows a simplified Modem library architecture:
-
-The application can adjust the size of these regions based on the needs to minimize the RAM requirements of the library.
-
+The application can adjust the size of these regions based on its requirements, with the exception of the Control region which has a fixed size that must be equal to ``NRF_MODEM_SHMEM_CTRL_SIZE``.
```
```suggestion
The application can adjust the size of these regions based on its requirements, with the exception of the Control region which has a fixed size that must be equal to the value of  ``NRF_MODEM_SHMEM_CTRL_SIZE``.
```

umapraseeda commented on nrf_modem/doc/architecture.rst:
```diff
@@ -57,6 +57,10 @@ The library OS abstraction layer defines the following functions to allocate and
 * :c:func:`nrf_modem_os_shm_tx_alloc`
 * :c:func:`nrf_modem_os_shm_tx_free`
 
+.. important:
+   It must be possible for the Modem library to allocate up to the number of bytes passed by the TX region size in the :c:struct:`nrf_modem_init_params` struct.
+   If the OS implementation of :c:func:`nrf_modem_os_shm_tx_alloc` require any overhead, additional memory must be set aside by the application.
```
```suggestion
.. important::
   It must be possible for the Modem library to allocate up to the number of bytes passed by the TX region size in the :c:struct:`nrf_modem_init_params` structure.
   If the OS implementation of :c:func:`nrf_modem_os_shm_tx_alloc` requires any overhead, additional memory must be set aside by the application.
```

umapraseeda commented on nrf_modem/doc/architecture.rst:
```diff
@@ -19,22 +19,22 @@ The following figure shows a simplified Modem library architecture:
+The application is responsible for reserving the memory area before passing it onto the library.
+
+In bootloader mode, the shared memory area must be as large as ``NRF_MODEM_SHMEM_BOOTLOADER_SIZE`` and its base address must be 4-bytes aligned.
+The library accepts the shared memory base address and size as parameters to the :c:func:`nrf_modem_bootloader_init` function through the :c:struct:`nrf_modem_bootloader_init_params`.
```
```suggestion
The library accepts the shared memory base address and size as parameters to the :c:func:`nrf_modem_bootloader_init` function through the :c:struct:`nrf_modem_bootloader_init_params` structure.
```

umapraseeda commented on nrf_modem/doc/architecture.rst:
```diff
@@ -19,22 +19,22 @@ The following figure shows a simplified Modem library architecture:
-
+The application can adjust the size of these regions based on its requirements, with the exception of the Control region which has a fixed size that must be equal to ``NRF_MODEM_SHMEM_CTRL_SIZE``.
+The base address of all regions must be 4-bytes aligned.
+The library accepts the layout of these regions as parameters to the :c:func:`nrf_modem_init` function through the :c:struct:`nrf_modem_shmem_cfg` struct in :c:struct:`nrf_modem_init_params`.
```
```suggestion
The library accepts the layout of these regions as parameters to the :c:func:`nrf_modem_init` function through the :c:struct:`nrf_modem_shmem_cfg` structure in :c:struct:`nrf_modem_init_params`.
```

umapraseeda commented on nrf_modem/doc/bootloader.rst:
```diff
@@ -7,74 +7,73 @@ Full firmware updates
 
-The modem library supports the full update of the modem firmware when the library is initialized in full DFU mode by the :file:`nrf_modem_full_dfu.h` interface.
-Full firmware update is the only operation that can be performed in full DFU mode.
+The modem library supports the full update of the modem firmware when the library is initialized in bootloader mode, enabling the :file:`nrf_modem_bootloader.h` interface.
```
```suggestion
Modem library supports the full update of the modem firmware when the library is initialized in bootloader mode, enabling the :file:`nrf_modem_bootloader.h` interface.
```

umapraseeda commented on nrf_modem/doc/bootloader.rst:
```diff
@@ -7,74 +7,73 @@ Full firmware updates
 The sockets are not available to the application.
 
-It is possible to switch between the full DFU and normal operation modes by reinitializing the library.
+It is possible to switch between the bootloader and normal operation modes by reinitializing the library, calling :c:func:`nrf_modem_shutdown` before :c:func:`nrf_modem_bootloader_init` or :c:func:`nrf_modem_init`, respectively.
```
```suggestion
It is possible to switch between the bootloader and normal operation modes by reinitializing the library, by calling :c:func:`nrf_modem_shutdown` before :c:func:`nrf_modem_bootloader_init` or :c:func:`nrf_modem_init`, respectively.
```

umapraseeda commented on nrf_modem/doc/bootloader.rst:
```diff
@@ -7,74 +7,73 @@ Full firmware updates
+The modem does not have a back-up ROM and hence, an interruption in the modem firmware update process can prevent the modem from further boot up.
 
-As a workaround to the above scenario, either the application must tolerate the situation or another means of recovery must be provided  (for example, reprogramming the modem with PC tools).
+As a workaround to the above scenario, either the application must tolerate the situation or another means of recovery must be provided, for example reprogramming the modem with PC tools.
```
```suggestion
As a workaround to the above scenario, either the application must tolerate the situation or another means of recovery must be provided, for example, reprogramming the modem with PC tools.
```

lemrey commented on nrf_modem/doc/CHANGELOG.rst:
```diff
@@ -9,6 +9,41 @@ Changelog
+  Use :c:func:`nrf_modem_bootloader_init` to initialize the library in bootloader mode.
+* Added a ``context`` parameter to :c:func:`nrf_modem_os_event_notify` to allow waking up only a subset of sleeping threads.
+* Added the :c:func:`nrf_modem_os_sleep` function.
+* The ``nrf_modem_limits.h`` file has been removed.
```
Would this work even if the file does not exist anymore?

umapraseeda replied:
Yes

umapraseeda replied:
This seems to be not fixed.

lemrey commented on nrf_modem/doc/logging.rst:
```diff
@@ -0,0 +1,27 @@
+
+The Modem library is released in two versions:
+* `libmodem.a`, without logs and
+* `libmodem_log.a` which is capable of emitting logs
```
I changed it to `printing logs`

lemrey commented:
@umapraseeda and @eivindj-nordic I have addressed your comments with [this](https://github.com/nrfconnect/sdk-nrfxlib/compare/3efa80e3048f6ba0b4bcf0c4b26a26332186b03f..23c1ba08f11d84706971478fe94de531bdb1770b) change.

eivindj-nordic commented on nrf_modem/include/nrf_modem_os.h:
```diff
@@ -107,9 +108,20 @@ int32_t nrf_modem_os_timedwait(uint32_t context, int32_t *timeout);
  *
- * This function shall wake all threads sleeping in @c nrf_modem_os_timedwait.
+ * This function wakes up all threads sleeping in @c nrf_modem_os_timedwait
+ * that have the same @p context, or all sleeping threads is @p context is zero.
```
```suggestion
 * that have the same @p context, or all sleeping threads if @p context is zero.
```

lemrey replied:
thanks, fixed

lemrey commented:
Have removed all `vsdx` files as requested by the tech writer.

umapraseeda commented on nrf_modem/doc/logging.rst:
```diff
@@ -0,0 +1,34 @@
+   :local:
+   :depth: 2
+
+The Modem library is released in two versions:
+* :file:`libmodem.a`, without logs
+* :file:`libmodem_log.a`, which is capable of printing logs
```
```suggestion
The Modem library is released in two versions:

* :file:`libmodem.a`, without logs
* :file:`libmodem_log.a`, which is capable of printing logs
```

umapraseeda commented on nrf_modem/doc/logging.rst:
```diff
@@ -0,0 +1,34 @@
+
+.. note::
+
+   The log version of the Modem library has an increased FLASH footprint.
```
```suggestion
   The log version of the Modem library has an increased flash footprint.
```

divipillai commented on nrf_modem/doc/ug_nrf_modem_porting_os.rst:
```diff
@@ -169,13 +169,13 @@ This function is called by the Modem library to put a thread to sleep unconditio
 
-* Start counting the time (this can be based on a Timer or Thread for instance).
-* In the case of a call to :c:func:`nrf_modem_os_event_notify` function, all waiting threads must wake up, reporting back the remaining time of the timer through the ``timeout`` variable.
+* Put the thread to sleep, if applicable, or otherwise wait until the amount time specified by ``timeout`` has elapsed, or until :c:func:`nrf_modem_os_event_notify` is called with the same ``context`` parameter, or with ``context`` equal to ``0``.
```
```suggestion
* Put the thread to sleep, if applicable, or otherwise wait until the amount of time specified by ``timeout`` has elapsed, or until :c:func:`nrf_modem_os_event_notify` is called with the same ``context`` parameter, or with ``context`` equal to ``0``
```

divipillai commented on nrf_modem/doc/ug_nrf_modem_porting_os.rst:
```diff
@@ -169,13 +169,13 @@ This function is called by the Modem library to put a thread to sleep unconditio
-* Start counting the time (this can be based on a Timer or Thread for instance).
-* In the case of a call to :c:func:`nrf_modem_os_event_notify` function, all waiting threads must wake up, reporting back the remaining time of the timer through the ``timeout`` variable.
+* Put the thread to sleep, if applicable, or otherwise wait until the amount time specified by ``timeout`` has elapsed, or until :c:func:`nrf_modem_os_event_notify` is called with the same ``context`` parameter, or with ``context`` equal to ``0``.
+* After the wait is complete, update the ``timeout`` parameter with the amount of time left to sleep. This shall be ``0`` if the timeout has elapsed or the amount of time left to sleep if the wait was interrupted by a call to :c:func:`nrf_modem_os_event_notify`.
```
```suggestion
* After the wait is complete, update the ``timeout`` parameter with the amount of time left to sleep. 
 This will be ``0`` if the timeout has elapsed or the amount of time left to sleep if the wait was interrupted by a call to :c:func:`nrf_modem_os_event_notify`.
```

lemrey commented:
Addressed last comments and squashed.

umapraseeda commented on nrf_modem/doc/bootloader.rst:
```diff
@@ -0,0 +1,145 @@
+The Modem library must then be manually initialized with the following code:
+
+.. code-block:: c
+
```
We could align the whole code block to the left similar to the other suggestions.
